### ISSUE TITLE
[FEATURE REQUEST] Correct stack trace in `request errored` pino-http's log

### ISSUE THREAD

**Issue by kasvith** [iamolegga/nestjs-pino#535]
**Is your feature request related to a problem? Please describe.**
By following #416 we now log the exception message. But it would be super helpful to log the actual stack trace there as well.
Most of the time we are unable to pinpoint the location without having the stack trace

**Describe the solution you'd like**
The library should automatically output the stack trace when handling HTTP errors

**Describe alternatives you've considered**

**Additional context**

**Comment by iamolegga** [comment]
thanks to @2zH for pointing out original reason. So the problem here is that `pino-http` doesn't play nice with `express` not passing the original error to logging middleware. Just checked this behavior with plain `express` and `pino-http`. Will check if something can be done with `pino-http` to fix that

**Comment by iamolegga** [comment]
A small update, stack should be fixed in the latest release for logs produced by nestjs built-in error handlers (more on this here #656). So in the current issue we are tracking stack of `request errored` log by`pino-http`

**Comment by antspk** [comment]
Correct me if I'm wrong or I am implementing this library in a wrong way. 

I've got this setup going on: 
```
LoggerModule.forRoot({
  exclude: ['/api/_health', '/_health', '/docs', '/api/docs'],
  pinoHttp: {
    genReqId: (msg: IncomingMessage): ReqId => ({
      trace: (msg.headers['x-trace-id'] as string) || v4(),
      span: v4(),
    }),
    logger: pino({
      base: {
        environment: process.env.APP_ENV || 'development',
        service: process.env.APP_NAME || 'unknown',
      },
    }),
  },
}),
```

And in my application I'm using custom errors, when an error is thrown like this:
```typescript
throw new InternalServerErrorException('Example');
```
I see this in logs:
```json
err: {
  "type": "Error",
  "message": "failed with status code 500",
  "stack":
      Error: failed with status code 500
          at ServerResponse.onResFinished (/Users/antanas/Projects/bourne-leisure/service-haven-holiday-bookings/node_modules/pino-http/logger.js:94:40)
          at ServerResponse.emit (node:events:402:35)
          at onFinish (node:_http_outgoing:830:10)
          at callback (node:internal/streams/writable:552:21)
          at afterWrite (node:internal/streams/writable:497:5)
          at afterWriteTick (node:internal/streams/writable:484:10)
          at processTicksAndRejections (node:internal/process/task_queues:82:21)
}
```
When instead I'd like to seem something like this:
```json
err: {
  "type": "InternalServerErrorException",
  "message": "Example",
  "stack":
      InternalServerErrorException: Example
          at SomeController.someMethod (/src/api/controllers/booking-controller.ts:60:11)
          at /node_modules/@nestjs/core/router/router-execution-context.js:38:29
          at processTicksAndRejections (node:internal/process/task_queues:96:5)
}
```

Now there's a logic in pino that does this here:
https://github.com/pinojs/pino-http/blob/a7abfb9dbe04d45ea05f1484af9ab6a556db71a6/logger.js#L93
and it was discussed here: https://github.com/pinojs/pino-http/issues/61

So idea here is that for pino to handle error properly it needs to have `err` property in the response object, which can be done with an intercetor:
```typescript
import { CallHandler, DynamicModule, ExecutionContext, NestInterceptor } from '@nestjs/common';
import { APP_INTERCEPTOR } from '@nestjs/core';
import { catchError, Observable, throwError } from 'rxjs';

export class NestJsPinoInterceptor implements NestInterceptor {
  intercept(context: ExecutionContext, next: CallHandler): Observable<any> | Promise<Observable<any>> {
    return next.handle().pipe(catchError((error) => throwError((context.switchToHttp().getResponse().err = error))));
  }
}
```
and then injected in module:
```
providers: [{ provide: APP_INTERCEPTOR, useClass: NestJsPinoInterceptor }],
```

If this is would an ok approach, it would be convenient if this library would provide this global interceptor since it's possible to have multiple global inteceptors:
https://github.com/nestjs/nest/blob/4b12f02f8c916e61a420841e7f0e3de0539961eb/packages/core/scanner.ts#L489

**Comment by iamolegga** [comment]
Thanks for researching @antspk, would you like to open a PR (with documentation and test) on this?

**Comment by iamolegga** [comment]
Or maybe I misunderstood your case. It looks like your logs are produced by error handler (or not? This is not full log, so I can not say for sure). In that case, it's already fixed in the latest version. If this is pino-http's `request errored` log then yes, we can add such interceptor. Anyway, I hope in the next couple of days I will find time to check that

**Comment by antspk** [comment]
I've created a minimal repository to show this issue: 
https://github.com/antspk/nestjs-pino-err/tree/main

With a solution in a separate branch:
https://github.com/antspk/nestjs-pino-err/tree/solution

If run something like this in both branches:
```shell
npm run start:dev | pino-pretty
```
and fire request e.g:
```shell
curl -X 'GET' http://localhost:3000
```

You should be able to see the difference in the `err` property, I would assume the behaviour with my suggestion should be default, and I'm not sure why it's not. I see that pino-http should handle it:

https://github.com/pinojs/pino-http/blob/a7abfb9dbe04d45ea05f1484af9ab6a556db71a6/logger.js#L85

as the `onResFinished` expect error to be passed in from `res.on('error', onResFinished)` callback, however this doesn't seem be be the case as the response `error` event doesn't seem to be fired, although `finish` event is fire, but if understand it properly it pass `err` as and argument. 

This might be an issue with express or nestjs it self not firing `error` event properly, or it is actually the way it supposed to be, I'm not sure about that.

However the solution to add additional interceptor which exposes `err` to response object seems to help.

**Comment by iamolegga** [comment]
Finally checked your solution. Let's add this to the library!

Can you make a new section in the docs which describes this feature, and add a test on this pretty same as [this one](https://github.com/iamolegga/nestjs-pino/blob/master/__tests__/errors.spec.ts#L176-L209). To make this work you'll need add a new method to the [TestCase class](https://github.com/iamolegga/nestjs-pino/blob/d737cbd621dd17fbf19234ec108bd219d5f5e8e4/__tests__/utils/test-case.ts#L17), something like `withInterceptor() { this.useInterceptor = true; return this }`. I think you'll figure it out, but feel free to tag me if you need help

**Comment by antspk** [comment]
Started looking into this and I think it would make sense to make this feature optional with additional parameter:

```typescript
/**
 * Optional parameter to add `err` property to the ServerResponse object,
 * that the `pino-http` module could pick it on print appropriate error.
 * The reason for this property is described the documentation:
 * @see https://github.com/iamolegga/nestjs-pino#faq.
 */
useInterceptor?: boolean;
```
Also if I understand this correctly, there there would not need to be a need to extend `TestCase` as the necessary configuration for the tests could be provided in a similar way to `useExisting`?

Then in the [LoggerModule](https://github.com/iamolegga/nestjs-pino/blob/fe7b119c310c1be9275c1ba6550f0376dc1f97e8/src/LoggerModule.ts#L34) additional condition could be which would add add, that additional provider to module exports:
```typescript
{
  provide: APP_INTERCEPTOR,
  useClass: LoggerErrorInterceptor,
}
```

However this would only for for `LoggerModule.forRoot` and not `LoggerModule.forRootAsync` as the async version uses factory method, which can't be used to change `DynamicModule` definition.

@iamolegga any ideas how we could tackle this? I wonder if it possible to somehow add interceptors dynamically when the module class is constructed.

**Comment by antspk** [comment]
Though about it a little bit more, and I think providing configuration option like this via provider, would be against nestjs design. We want to configure module not a dependency so it the configuration options should come from a top level argument in a module.

Anyway I guess it just means that `LoggerModuleAsyncParams` will need slight modification as well.

**Comment by iamolegga** [comment]
Sorry, I didn't get it. Why `{provide: APP_INTERCEPTOR, useClass: LoggerErrorInterceptor}` can not be exported from the module along with other providers for async configuration? Have you checked it yet? If so it sounds like a limitation that should be fixed in the framework 😀 But in that case let's just ship this interceptor. When there won't be any limitation for `useInterceptor` field in the config we can add it there too. A man can do no more than he can)

**Comment by iamolegga** [comment]
> Anyway I guess it just means that LoggerModuleAsyncParams will need slight modification as well

Let's keep sync and async params the same. If `useInterceptor` is not possible for both of them, then let's just export it from the package and that's it

**Comment by iamolegga** [comment]
@antspk hi, will you be able to open PR? Or should I just take your fix and finish all the rest?

**Comment by antspk** [comment]
Sorry I've been really busy with other stuff. I've made a PR: https://github.com/iamolegga/nestjs-pino/pull/722, but it needs some fixing, as the interceptor doesn't seem to work with fastify adapter. Maybe you could look into it?

**Comment by iamolegga** [comment]
@antspk no problem. Thanks for investing your time in this

> Maybe you could look into it?

will do in a next couple of days

**Comment by gwajanara** [comment]
@antspk @iamolegga 
In case fastify, try response.raw.err = err

const response = ctx.getResponse<FastifyReply>();
response.err = err // X
response.raw.err = err // O

**Comment by iamolegga** [comment]
fixed in `2.5.0` with [LoggerErrorInterceptor](https://github.com/iamolegga/nestjs-pino#expose-stack-trace-and-error-class-in-err-property). Thanks to @antspk
